File: OJ/2020/200405/LC122.java
/**
 * LeetCode 122 Easy
 * Best Time to Buy and Sell Stock II
 * Dynamic Programming
 */

public class LC122 {
    public static void main(String [] args){
        Solution solution = new Solution();
        System.out.println(solution.maxProfit(new int [] {7,1,5,3,6,4}));
        System.out.println(solution.maxProfit(new int [] {1,2,3,4,5}));
    }

    static
    class Solution {
        public int maxProfit(int[] prices) {
            if (prices.length < 2) {
                return 0;
            }
            int maxHold = -prices[0];
            int maxSold = 0;
            for (int i = 1; i < prices.length; i++) {
                int hold = Math.max(maxHold, maxSold - prices[i]);
                int sold = Math.max(maxSold, maxHold + prices[i]);
                maxHold = hold;
                maxSold = sold;
            }
            return maxSold;
        }
    }
}

File: OJ/2021/210720/LC122.py
# LeetCode 122
# Best Time to Buy and Sell Stock II
# Dynamic Programming

from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) < 2:
            return 0
        max_buy, max_sell = - prices[0], 0
        for i in range(1, len(prices)):
            max_buy, max_sell = max(max_buy, max_sell - prices[i]), max(max_sell, prices[i] + max_buy)
        return max_sell

if __name__ == "__main__":
    print(Solution().maxProfit([7,1,5,3,6,4]))
    print(Solution().maxProfit([1,2,3,4,5]))
    print(Solution().maxProfit([7,6,4,3,1]))